Core Concept week 1 day 6 update (#41)
microsoft/win-student-devs: docs/30DaysOfPWA/core-concepts/06.md

| _Exercise_ | Explore capabilities support in sample PWA. |
| _Related_ | [Week 2: Advanced Capabilities](../advanced-capabilities/) |

---

## Let's Recap

We've learned that:
 * Progressive Web Apps use _progressive enhancement_ to deliver native-like experiences on capable devices, while falling back to familiar, usable website behaviors elsewhere.
 * Implementation uses _open web technologies_ that enable behaviors like installability, network-independent operation, safety, and discoverability (in search and app stores).
 * Core building blocks are _HTTPS_ (safe operation), _Service Workers_ (work offline, re-engage users), and _Web App Manifests_ (install on device, publish to store). 
 * Service Workers are the beating heart of PWAs. We learned about storage options, caching strategies and lifecycle events that help us implement key functionality for offline-ready app experiences.

But wait, there's more. Platforms keep evolving rapidly, and PWAs need to keep up with their platform-native counterparts in capabilities. Today we'll look at the final piece of the puzzle: _web capabilities_.

---

## What are Web Capabilities?

> Let's start with the motivation. Device platforms today evolve faster than the web platform can. Native apps can keep pace, delivering richer experiences driven by new device hardware or features. But web apps are limited by their ability follow suit because they need those capabilities to work _consistently across all device platforms_ that browsers may run on. This creates the _experience gap_ between web and platform-native app experiences that progressive web apps try to fill.

Reviewer: If a section header is a question, I usually expect the first sentence to answer the question right away. Maybe give a one-liner and then get into depth/motivation?

> Let's start with the motivation. Device platforms today evolve faster than the web platform can. Native apps can keep pace, delivering richer experiences driven by new device hardware or features. But web apps are limited by their ability follow suit because they need those capabilities to work _consistently across all device platforms_ that browsers may run on. This creates the _experience gap_ between web and platform-native app experiences that progressive web apps try to fill.

Reviewer 2: ```suggestion
Let's start with the motivation. Device platforms today evolve faster than the web platform can. Native apps can keep pace, delivering richer experiences driven by new device hardware or features. But web apps are limited in their ability follow suit because they need those capabilities to work _consistently across all device platforms_ that browsers may run on. This creates the _experience gap_ between web and platform-native app experiences that progressive web apps try to fill.
```

Reviewer 2: Also, this is not totally accurate:

> they need those capabilities to work _consistently across all device platforms_ that browsers may run on

Often you’ll see browsers, especially those linked to an OS, roll out features for the web in advance of there being a standard (or even standards track idea) backing them. Without wading into those waters, however, you could say something like this, which is always accurate:

> they need those capabilities to be implemented for native software first so the browser can then proxy them into a web platform feature.

So, what if those device-specific capabilities could be exposed to the web platform through _standardized interfaces_ that all stakeholders agree on? Now, web apps gain access to the same device capabilities as platform-native apps, and they do this without losing their core web identity and characteristics.

> This is the promise of the [Web Capabilities Project](https://aka.ms/learn-PWA/30Days-1.6/developers.google.com/web/updates/capabilities) - now known as _Project Fugu_. 

Reviewer 2: ```suggestion
This is the promise of the [Web Capabilities Project](https://aka.ms/learn-PWA/30Days-1.6/developers.google.com/web/updates/capabilities), also known as _Project Fugu_. 
```

---

## What is Project Fugu?

> [Project Fugu](https://aka.ms/learn-PWA/30Days-1.6/www.chromium.org/teams/web-capabilities-fugu) is a collaborative initiative with one driving vision: _"web apps should be able to do anything that platform-native apps can"_. Contributors include representatives from Microsoft, Intel, Samsung, Google - and other stakeholders vested in the success of the web platform.

Reviewer 2: ```suggestion
[Project Fugu](https://aka.ms/learn-PWA/30Days-1.6/www.chromium.org/teams/web-capabilities-fugu) is a collaborative initiative with one driving vision: _"web apps should be able to do anything that platform-native apps can"_. Contributors include representatives from Google, Intel, Microsoft, Samsung, and other stakeholders vested in the success of the web platform.
```

Putting them alphabetically to avoid playing favorites 😄

> Why Fugu? It's the Japanese word for _pufferfish_ (also the project logo) that is a desired food, but that can kill you if prepared incorrectly. [This article](https://aka.ms/learn-PWA/30Days-1.6/felixgerschau.com/web-capabilities-project-fugu-google) explains that the Chromium team chose it to reflect their deep belief that _if_ they violated the core tenets of the web (user security, privacy, trust) it would adversely impact the project's ability to thrive.

Reviewer 2: ```suggestion
Why Fugu? It's the Japanese word for _pufferfish_ (also the project logo) that is a delicacy, but can kill you if prepared incorrectly. [This article](https://aka.ms/learn-PWA/30Days-1.6/felixgerschau.com/web-capabilities-project-fugu-google) explains that the Chromium team chose it to reflect their deep belief that _if_ they violated the core tenets of the web (user security, privacy, trust) it would adversely impact the project's ability to thrive.
```

How does Fugu work? It's a multi-stage process where proposals for new capability APIs are made in the open and evolved iteratively with feedback from all key stakeholders. The steps of the [standardization process](https://aka.ms/learn-PWA/30Days-1.6/web.dev/fugu-status) are as follows:

 1. New capability is suggested - by filing a [feature request](https://aka.ms/learn-PWA/30Days-1.6/web.dev/fugu-status)
 2. Capability goes [under consideration](https://aka.ms/learn-PWA/30Days-1.6/fugu-tracker.web.app#under-consideration) - explainers written, feedback solicited.
 3. Work has [started](https://aka.ms/learn-PWA/30Days-1.6/fugu-tracker.web.app#started) - on a formal API spec to iterate on,
 4. Available in [developer trials](https://aka.ms/learn-PWA/30Days-1.6/fugu-tracker.web.app#developer-trial) - under experimental flag, not for production
 5. Available in [origin trials](https://aka.ms/learn-PWA/30Days-1.6/fugu-tracker.web.app#origin-trial) - for validation, in prototypes for larger audiences
 6. Capability [has shipped](https://aka.ms/learn-PWA/30Days-1.6/fugu-tracker.web.app#shipped) - API is _stable_ in latest browser version.

> The [Fugu Tracker](https://aka.ms/learn-PWA/30Days-1.6/fugu-tracker.web.app) provides real-time updates about each capability API's status in this pipeline.

Author: Adding @aarongustafson as 2.1 is about web capabilities intro.

---

## What capabilities can I use today?

> Look through the [Shipped](https://aka.ms/learn-PWA/30Days-1.6/fugu-tracker.web.app#shipped) list on the Project Fugu API tracker, and click on an API to learn what it can do _and how well it is supported on various browsers_ (including on mobile Android and iOS platforms). Here are some examples from that list.

Reviewer 2: ```suggestion
Look through [the Shipped list on the Project Fugu API tracker](https://aka.ms/learn-PWA/30Days-1.6/fugu-tracker.web.app#shipped) and click on an API to learn what it can do _and how well it is supported on various browsers_ (including on mobile Android and iOS platforms). Here are some examples from that list:
```

 * Project Fugu provides an API tracker to learn about current capability proposals and their status in the standards pipeline.
 * Capability proposals are handled in the open, with status going from feature request, to under consideration, started, dev and origin trials - before final shipping.
 * You can explore sample PWAs to understand how these capabilities are used (experience) and integrated (code).

And don't forget to look out for a full week of deep dives into advanced capabilities just days away!

---

## Learning Resources

 1. [Unlocking new Capabilities for the web](https://aka.ms/learn-PWA/30Days-1.6/developers.google.com/web/updates/capabilities#process) - the process.
 2. [Project Fugu: Status](https://aka.ms/learn-PWA/30Days-1.6/web.dev/fugu-status) - the blog
 3. [Project Fugu: Tracker](https://aka.ms/learn-PWA/30Days-1.6/fugu-tracker.web.app#shipped) - the dashboard
 4. [Project Fugu: Chromium](https://bugs.chromium.org/p/chromium/issues/list?can=2&q=proj-fugu) - the proposals
 5. [Sample PWAs](https://aka.ms/learn-PWA/30Days-1.6/docs.microsoft.com/en-us/microsoft-edge/progressive-web-apps-chromium/demo-pwas) - Microsoft Edge capabilities

>  6. [Experimental Features + Origin Trials](https://aka.ms/learn-PWA/30Days-1.6/docs.microsoft.com/en-us/microsoft-edge/progressive-web-apps-chromium/how-to/origin-trials) - Microsoft Edge

Author: I like that exercise is at the end. Keep resources as part of the reading.
